transports/onion/src/address.rs

```diff
@@ -0,0 +1,107 @@
+use std::net::{IpAddr, SocketAddr};
+
+use arti_client::{IntoTorAddr, TorAddr, TorAddrError};
+use libp2p_core::{multiaddr::Protocol, Multiaddr};
+
+pub fn dangerous_extract_tor_address(multiaddr: &Multiaddr) -> Result<TorAddr, TorAddrError> {
+    let socket_addr = try_extract_socket_addr(&mut multiaddr.clone())?;
+    TorAddr::dangerously_from(socket_addr)
+}
+
+pub fn safe_extract_tor_address(multiaddr: &mut Multiaddr) -> Result<TorAddr, TorAddrError> {
```

Comment:
To align the signatures, this should probably take an owned version too.

Reply:
I could do that but that might be incorrect because, as you said:
> I'd suggest we instead pop off two items from the iterator, ignore the rest and match the first one for any of the DNS variants and the second one on `Tcp`. If they are anything else, bail with `MultiAddrNotSuppored` (this is currently done wrong in the `dial` implementation which means the next transport will not get selected in a chain of transports).
If it's not mutable we don't get the changes to the multiaddress. However I think there is an easy solution.

Reply:
Nevermind, I don't see an ideal solution. If we want to make the pop operations count i consider `&mut` a way better solution. I tried it with owned versions, however this would require way more clones. If you have a better idea, I would be glad.

Reply:
I've pushed 8c20c301, let me know what you think :)